src/Symfony/Component/Intl/Data/Generator/ZoneDataGenerator.php

```diff
@@ -0,0 +1,172 @@
+                ),
+            );
+
+            $this->zoneCodes = array_merge($this->zoneCodes, array_keys($data['Names']));
+
+            return $data;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function generateDataForRoot(BundleReaderInterface $reader, $tempDir)
+    {
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function generateDataForMeta(BundleReaderInterface $reader, $tempDir)
+    {
+        $rootBundle = $reader->read($tempDir, 'root');
+
+        $this->zoneCodes = array_unique($this->zoneCodes);
+
+        sort($this->zoneCodes);
+
+        $data = array(
+            'Version' => $rootBundle['Version'],
+            'Zones' => $this->zoneCodes,
+        );
+
+        return $data;
+    }
+
+    private static function generateZones(ArrayAccessibleResourceBundle $typeBundle, ArrayAccessibleResourceBundle $metaBundle, ArrayAccessibleResourceBundle $rootBundle, ArrayAccessibleResourceBundle $localeBundle): array
+    {
+        $available = array();
+        foreach ($typeBundle['typeMap']['timezone'] as $zone => $_) {
+            if (self::UNKNOWN_ZONE === $zone) {
+                continue;
+            }
+
+            $available[$zone] = true;
+        }
+
+        $metazones = array();
+        foreach ($metaBundle['metazoneInfo'] as $zone => $info) {
+            foreach ($info as $metazone) {
+                $metazones[$zone] = $metazone->get(0);
+            }
+        }
+
+        $zones = array();
+        foreach (array_keys($available) as $zone) {
+            $name = null;
+            $city = null;
+
+            if (isset($localeBundle['zoneStrings'][$zone])) {
+                $data = $localeBundle['zoneStrings'][$zone];
+                $name = $data->get('ls') ?? $data->get('lg') ?? $data->get('ld');
+                $city = $data->get('ec');
+            } elseif (isset($rootBundle['zoneStrings'][$zone])) {
+                $data = $rootBundle['zoneStrings'][$zone];
+                $name = $data->get('ls') ?? $data->get('lg') ?? $data->get('ld');
+                $city = $data->get('ec');
+            }
+
+            if (null === $city && 0 !== strrpos($zone, 'Etc:') && false !== $i = strrpos($zone, ':')) {
+                $city = str_replace('_', ' ', substr($zone, $i + 1));
+            }
+
+            if (null === $name && isset($metazones[$zone]) && isset($localeBundle['zoneStrings']['meta:'.$metazones[$zone]])) {
+                $metadata = $localeBundle['zoneStrings']['meta:'.$metazones[$zone]];
+                $name = $metadata->get('ls') ?? $metadata->get('lg') ?? $metadata->get('ld') ?? $name;
+            }
+
+            if (null !== $city) {
+                if (null === $name) {
+                    $name = str_replace('{0}', $city, $localeBundle['zoneStrings']['regionFormat']);
+                } else {
+                    $name .= ' ('.$city.')';
+                }
+            }
+
+            if (null === $name || '' === $name) {
```

Comment:
double check where the empty string comes from. Also if name is null it should use the zone ID instead.